### PR TITLE
Improved error message for invalid private row type declaration

### PR DESCRIPTION
**Pull request by Octachron**
This small PR proposes to remove the `Bad_fixed_type` error constructor. Those errors were raised when trying to make private a type without a row variable in 3.10
```OCaml
type t = private int
```
> This fixed type is not an object or variant

or
```OCaml
type t = private [ `A ]
```
> Error: This fixed type has no row variable

Nowadays both cases are allowed (since 4.02) and the error cannot be raised anymore: all uses of `Typedecl.set_fixed_row` (the only function that raises such errors) are guarded by call to `Typedecl.is_fixed_type`  (see https://github.com/Octachron/ocaml/commit/ff4bb0fd676f9fca628136c14c279c18358f7b38 for a version of this PR wich makes this point more salient) which precludes the error branches in `Typedecl.set_fixed_row`.

### CODE REVIEW

File: typing/typedecl.mli
Comment (trefis) [inline typing/typedecl.mli]: What would you think of switching to an optional argument, for instance `?fixed_row_path:Path.t`?

**Review by trefis — APPROVED** [review]
This seems fine.

File: typing/typedecl.mli
Comment (Octachron) [reply typing/typedecl.mli]: You mean using a meaningful label for an optional argument used once in the whole code base? I buy it.

**Comment by trefis** [comment]
You apparently need to rebase, but after that, feel free to merge as soon as the CI is green.

**Comment by Octachron** [comment]
Rechecking the code, there is one way to raise a `no fixed row` error, but it looks like a bug in itself:
```OCaml
module M: sig type t = private [< `A > `A ] = struct type  t= [`A ] end
``` 
raises
>Error: This fixed type has no row variable

because the fact that the explicit lower bounds match the upper bound is seen to late, whereas the more sensible
```OCaml
module M: sig type t = private [ `A ] = struct type  t= [`A ] end
``` 
is accepted.

I will update this PR to fix this path.

**Comment by Octachron** [comment]
After some discussion with @garrigue , it appears that we want to keep half of the `Bad_fixed_type_error` cases. There are the case where the user had written something that looks like it has a row type variable that could be made private, but in fact there is no such free row variable. The error message now looks like:
```OCaml
type a = [`A | `C | `D]
type b = [`B | `D | `E]
type c = private [< a | b > `A `B `C `D `E]
```
```OCaml
Error: This private row type declaration is invalid.
       The type expression on the right-hand side reduces to
         [ `A | `B | `C | `D | `E ]
       which does not have a free row type variable.
       Hint: If you intended to define a private type abbreviation,
       write explicitly
         private [ `A | `B | `C | `D | `E ]
```

**Comment by gasche** [comment]
This would need to be rebased, and then a new approval by @garrigue I guess.

**Comment by Octachron** [comment]
I have rebased this PR (and cleaned the history of the initial removal). Indeed, it would need an approval for the expanded  error message.

File: typing/typedecl.ml
Comment (gasche) [inline typing/typedecl.ml:188]: The comment here indicates that we know something about the user intent ("the syntax"), but the comment on the function declaration just says that "is_fixed_type" is a precondition. Maybe the comment on the function declaration could also, for people who are not expert on what "is_fixed_type" means, point out that we are working with a private type declaration?

**Review by gasche — APPROVED** [review]
This looks good to me.

File: typing/typedecl.ml
Comment (Octachron) [reply typing/typedecl.ml:188]: Good point. I would propose to update the name of the function to `set_private_row` too.

Comment (gasche) [reply typing/typedecl.ml:188]: why not, but many places use `fixed` as an adjective, should they be updated too?

Comment (Octachron) [reply typing/typedecl.ml:188]: A private row is a subcase of fixed rows, but `Typedecl` and `set_fixed_row` only generate this `Fixed_private` subcase.

**Comment by gasche** [comment]
(We need @Octachron to rebase and possibly do changes due to the nitpick, and then this should be good to merge.)

**Pull request merged**
